assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java

```diff
@@ -4656,17 +4655,31 @@ private void notificationToInstructors(AssignmentSubmission submission, Assignme
         }
     }
 
-    private void notificationToStudent(AssignmentSubmission submission) {
+    private void notificationToStudent(AssignmentSubmission submission, Assignment assignment) {
         if (serverConfigurationService.getBoolean("assignment.submission.confirmation.email", true)) {
-            Set<String> submitterIds = submission.getSubmitters().stream().map(AssignmentSubmissionSubmitter::getSubmitter).collect(Collectors.toSet());
-            Set<User> users = submitterIds.stream().map(id -> {
-                try {
-                    return userDirectoryService.getUser(id);
-                } catch (UserNotDefinedException e) {
-                    log.warn("Could not find user with id = {}, {}", id, e.getMessage());
-                }
-                return null;
-            }).filter(Objects::nonNull).collect(Collectors.toSet());
+            String siteId = assignment.getContext();
+            Set<String> siteUsers = null;
+            try {
+                siteUsers = siteService.getSite(siteId).getUsers();
+            } catch (IdUnusedException e) {
+                log.warn("Could not find site: {}", siteId, e);
+            }
+            Set<User> users = submission.getSubmitters()
+                    .stream()
+                    .map(AssignmentSubmissionSubmitter::getSubmitter)
+                    .filter(siteUsers::contains) // this filters active users
+                    .collect(Collectors.toSet())
```

Comment:
Is collecting here necessary?

Reply:
yeah i tried to remove it

Reply:
That's odd, filter() just returns a stream anyway. My IDE doesn't complain if I remove the collect.

Reply:
ahhh remove collect and the stream.... ok done